File: backend/src/api/validators.ts
import type {
  ActionKind,
  ActionRequest,
  Direction,
  Orientation,
  ShipKind,
} from "@shared/index.js";
import { ApiError } from "./errors.js";

const SHIP_KINDS: ShipKind[] = [
  "AIRCRAFT_CARRIER",
  "BATTLESHIP",
  "CRUISER",
  "FRIGATE",
  "SUBMARINE",
];

const ACTION_KINDS: ActionKind[] = [
  "SINGLE_HIT",
  "AREA_HIT_2X2",
  "MOVE_1",
  "ROTATE_90",
];

const ORIENTATIONS: Orientation[] = ["horizontal", "vertical"];

const DIRECTIONS: Direction[] = ["N", "E", "S", "W"];

function need(cond: unknown, msg: string): asserts cond {
  if (!cond) throw new ApiError("BAD_REQUEST", msg);
}

export function asString(v: unknown, field: string): string {
  need(typeof v === "string" && v.length > 0, `Missing field: ${field}`);
  return v as string;
}

export function asNonNegInt(v: unknown, field: string): number {
  need(typeof v === "number" && Number.isInteger(v) && v >= 0, `Field ${field} must be a non-negative integer`);
  return v as number;
}

export function asShipKind(v: unknown): ShipKind {
  need(typeof v === "string" && SHIP_KINDS.includes(v as ShipKind), `Invalid ship kind: ${v}`);
  return v as ShipKind;
}

export function asOrientation(v: unknown): Orientation {
  need(typeof v === "string" && ORIENTATIONS.includes(v as Orientation), `Invalid orientation: ${v}`);
  return v as Orientation;
}

export function asActionKind(v: unknown): ActionKind {
  need(typeof v === "string" && ACTION_KINDS.includes(v as ActionKind), `Invalid action kind: ${v}`);
  return v as ActionKind;
}

export function asDirection(v: unknown): Direction {
  need(typeof v === "string" && DIRECTIONS.includes(v as Direction), `Invalid direction: ${v}`);
  return v as Direction;
}

/** Validate the body of `POST /api/games/:id/actions`. */
export function parseActionRequest(body: unknown): ActionRequest {
  need(body && typeof body === "object", "Request body must be a JSON object");
  const obj = body as Record<string, unknown>;

  if (obj.endTurn === true) {
    return { endTurn: true };
  }

  const shipId = asString(obj.shipId, "shipId");
  const kind = asActionKind(obj.kind);
  const targetsRaw = obj.targets;
  need(Array.isArray(targetsRaw), "Field 'targets' must be an array of cell indices");
  const targets = (targetsRaw as unknown[]).map((t, i) => asNonNegInt(t, `targets[${i}]`));
  const out: ActionRequest = { shipId, kind, targets };
  if (obj.direction !== undefined) {
    (out as { direction?: Direction }).direction = asDirection(obj.direction);
  }
  return out;
}

/** Validate `POST /api/games/:id/place`. */
export function parsePlaceBody(body: unknown): {
  kind: ShipKind;
  anchor: number;
  orientation: Orientation;
} {
  need(body && typeof body === "object", "Request body must be a JSON object");
  const obj = body as Record<string, unknown>;
  return {
    kind: asShipKind(obj.kind),
    anchor: asNonNegInt(obj.anchor, "anchor"),
    orientation: asOrientation(obj.orientation),
  };
}

/** Validate `POST /api/games` body (optional seed). */
export function parseCreateBody(body: unknown): { seed?: number } {
  if (!body || typeof body !== "object") return {};
  const obj = body as Record<string, unknown>;
  if (obj.seed === undefined) return {};
  return { seed: asNonNegInt(obj.seed, "seed") };
}

File: backend-rs/src/api/validators.rs
//! Manual parsers that mirror `backend/src/api/validators.ts`. We use
//! `serde_json::Value` to keep the validation messages identical to the TS
//! implementation, which the frontend may depend on for error display.

use serde_json::Value;

use crate::shared::{
    ActionKind, ActionRequest, Direction, Orientation, ShipActionRequest, ShipKind,
};

use super::errors::{ApiError, ApiErrorCode};

fn bad<M: Into<String>>(msg: M) -> ApiError {
    ApiError::new(ApiErrorCode::BadRequest, msg)
}

fn need(cond: bool, msg: &str) -> Result<(), ApiError> {
    if !cond {
        Err(bad(msg.to_string()))
    } else {
        Ok(())
    }
}

pub fn as_string(v: &Value, field: &str) -> Result<String, ApiError> {
    if let Value::String(s) = v {
        if !s.is_empty() {
            return Ok(s.clone());
        }
    }
    Err(bad(format!("Missing field: {}", field)))
}

pub fn as_non_neg_int(v: &Value, field: &str) -> Result<u32, ApiError> {
    if let Value::Number(n) = v {
        if let Some(i) = n.as_i64() {
            if i >= 0 && i <= u32::MAX as i64 {
                return Ok(i as u32);
            }
        }
        // Some JSON numbers might come through as floats; reject non-integers.
    }
    Err(bad(format!(
        "Field {} must be a non-negative integer",
        field
    )))
}

pub fn as_ship_kind(v: &Value) -> Result<ShipKind, ApiError> {
    let s = match v {
        Value::String(s) => s.as_str(),
        _ => "",
    };
    match s {
        "AIRCRAFT_CARRIER" => Ok(ShipKind::AircraftCarrier),
        "BATTLESHIP" => Ok(ShipKind::Battleship),
        "CRUISER" => Ok(ShipKind::Cruiser),
        "FRIGATE" => Ok(ShipKind::Frigate),
        "SUBMARINE" => Ok(ShipKind::Submarine),
        _ => Err(bad(format!("Invalid ship kind: {}", s))),
    }
}

pub fn as_orientation(v: &Value) -> Result<Orientation, ApiError> {
    let s = match v {
        Value::String(s) => s.as_str(),
        _ => "",
    };
    match s {
        "horizontal" => Ok(Orientation::Horizontal),
        "vertical" => Ok(Orientation::Vertical),
        _ => Err(bad(format!("Invalid orientation: {}", s))),
    }
}

pub fn as_action_kind(v: &Value) -> Result<ActionKind, ApiError> {
    let s = match v {
        Value::String(s) => s.as_str(),
        _ => "",
    };
    match s {
        "SINGLE_HIT" => Ok(ActionKind::SingleHit),
        "AREA_HIT_2X2" => Ok(ActionKind::AreaHit2x2),
        "MOVE_1" => Ok(ActionKind::Move1),
        "ROTATE_90" => Ok(ActionKind::Rotate90),
        _ => Err(bad(format!("Invalid action kind: {}", s))),
    }
}

pub fn as_direction(v: &Value) -> Result<Direction, ApiError> {
    let s = match v {
        Value::String(s) => s.as_str(),
        _ => "",
    };
    match s {
        "N" => Ok(Direction::N),
        "E" => Ok(Direction::E),
        "S" => Ok(Direction::S),
        "W" => Ok(Direction::W),
        _ => Err(bad(format!("Invalid direction: {}", s))),
    }
}

/// Validate the body of `POST /api/games/:id/actions`.
pub fn parse_action_request(body: &Value) -> Result<ActionRequest, ApiError> {
    need(
        body.is_object(),
        "Request body must be a JSON object",
    )?;
    let obj = body.as_object().unwrap();

    if let Some(et) = obj.get("endTurn") {
        if et == &Value::Bool(true) {
            return Ok(ActionRequest::EndTurn);
        }
    }

    let ship_id = match obj.get("shipId") {
        Some(v) => as_string(v, "shipId")?,
        None => return Err(bad("Missing field: shipId")),
    };
    let kind = match obj.get("kind") {
        Some(v) => as_action_kind(v)?,
        None => return Err(bad("Invalid action kind: undefined")),
    };
    let targets_v = obj
        .get("targets")
        .ok_or_else(|| bad("Field 'targets' must be an array of cell indices"))?;
    let targets_arr = targets_v
        .as_array()
        .ok_or_else(|| bad("Field 'targets' must be an array of cell indices"))?;
    let mut targets: Vec<u32> = Vec::with_capacity(targets_arr.len());
    for (i, t) in targets_arr.iter().enumerate() {
        targets.push(as_non_neg_int(t, &format!("targets[{}]", i))?);
    }
    let mut out = ShipActionRequest {
        ship_id,
        kind,
        targets,
        direction: None,
    };
    if let Some(d) = obj.get("direction") {
        if !d.is_null() {
            out.direction = Some(as_direction(d)?);
        }
    }
    Ok(ActionRequest::Ship(out))
}

pub fn parse_place_body(
    body: &Value,
) -> Result<(ShipKind, u32, Orientation), ApiError> {
    need(
        body.is_object(),
        "Request body must be a JSON object",
    )?;
    let obj = body.as_object().unwrap();
    let kind = as_ship_kind(
        obj.get("kind")
            .ok_or_else(|| bad("Invalid ship kind: undefined"))?,
    )?;
    let anchor = as_non_neg_int(
        obj.get("anchor")
            .ok_or_else(|| bad("Field anchor must be a non-negative integer"))?,
        "anchor",
    )?;
    let orientation = as_orientation(
        obj.get("orientation")
            .ok_or_else(|| bad("Invalid orientation: undefined"))?,
    )?;
    Ok((kind, anchor, orientation))
}

pub fn parse_create_body(body: &Value) -> Result<Option<u32>, ApiError> {
    if !body.is_object() {
        return Ok(None);
    }
    let obj = body.as_object().unwrap();
    let Some(s) = obj.get("seed") else {
        return Ok(None);
    };
    if s.is_null() {
        return Ok(None);
    }
    Ok(Some(as_non_neg_int(s, "seed")?))
}
